openai_api.py

```diff
@@ -0,0 +1,596 @@
+                            inputs=token_ids,
+                            max_new_tokens=max_new_tokens - total_generated_tokens,
+                            # Adjust max_new_tokens based on tokens already generated
+                            session=session,
+                        )
+                        delta = outputs[0, n_input_tokens:].tolist()
+                        total_generated_tokens += len(delta)
+
+                        logger.info(f"Generated {len(delta)} tokens")
+                        logger.info(f"Tokens generated: delta_q={repr(delta_q)}, delta={repr(delta)}")
+
+                        outputs_text = utils.safe_decode(tokenizer, delta_q + delta)
+
+                        if outputs_text[-10:].find("\ufffd") > -1:
+                            # If there's a replacement character, keep getting more tokens
+                            # until we can decode properly
+                            delta_q += delta
+                            logger.info(f"Retry generation {repr(outputs_text)}")
+                        else:
+                            response_json = create_stream_response_json(index=0, text=outputs_text)
+                            logger.info(f"Generated {repr(outputs_text)}")
+                            yield f"data: {response_json}\n\n"
+                            delta_q = []  # Clear delta_q for the next iteration
+                            token_ids = None  # Clear inputs for the next generation step
+
+                        if len(delta) > 0 and delta[-1] == tokenizer.eos_token_id:
+                            yield END_STRING
+                            return
+
+                yield END_STRING
+
+        return StreamingResponse(completion_stream_generator(model, token_ids),
+                                 media_type="text/event-stream")
+
+    num_prompt_tokens = token_ids.shape[1]
+
+    results_generator = model.generate(token_ids, **generation_config.to_diff_dict())
+
+    choices = []
+    for index, output_tokens in enumerate(results_generator):
+        text = tokenizer.decode(
+            output_tokens[num_prompt_tokens:],
+            skip_special_tokens=skip_special_tokens,
+            spaces_between_special_tokens=spaces_between_special_tokens,
+        )
+
+        choice_data = ChatCompletionResponseChoice(
+            index=index,
+            message=ChatMessage(role="assistant", content=text),
+            finish_reason=None,
+        )
+        choices.append(choice_data)
+
+    num_generated_tokens = sum(
+        len(output) for output in results_generator)
+    usage = UsageInfo(
+        prompt_tokens=num_prompt_tokens,
+        completion_tokens=num_generated_tokens,
+        total_tokens=num_prompt_tokens + num_generated_tokens,
+    )
+    response = ChatCompletionResponse(
+        id=request_id,
+        created=created_time,
+        model=model_name,
+        choices=choices,
+        usage=usage,
+    )
+
+    if request.stream:
```

Comment:
May be it's better to send unprocesseble entity? How openai answering when you choose choices with streaming?

Reply:
Because of OpenAI documentation, it is mandatory to send data like streaming in any case (the stream terminated by a `data: [DONE]` message).